gpii/node_modules/gpii-iniFile/index.js

```diff
@@ -0,0 +1,512 @@
+ *
+ * @callback iniFile.gotValue
+ * @param {String} sectionPath The path of the section.
+ * @param {String} key The value name.
+ * @param {String} value The value.
+ * @param {Boolean} quoted true if the value was quoted.
+ * @return {String} The new value, fluid.NO_VALUE to remove the value, or undefined if it's unchanged. Ignored when
+ *  reading.
+ */
+
+/**
+ * Parse an INI file, invoking call backs at interesting points: start/end of a section, and on a key=value.
+ *
+ * If options.write=true, the return values of the callbacks will be used to modify the content, the entirety of which
+ * is then returned by this function.
+ *
+ * It uses the regular expression in ./ini.regex to perform the actual parsing.
+ *
+ * @param {String} content The INI file content.
+ * @param {Object} handlers Object containing the callbacks.
+ * @param {iniFile.sectionBegin} handlers.sectionBegin - called at the start of a new section, returning a string to
+ *  write before the section line, if options.write=true.
+ * @param {iniFile.sectionEnd} handlers.sectionEnd - called at the end of a section, after all of its sub-sections
+ *  have been seen. Returning a string to write at the end of the section, if options.write=true.
+ * @param {iniFile.gotValue} handlers.gotValue - called when a value has been parsed. If options.write=true, return the
+ * text to replace the content of the value, undefined for no update, or fluid.NO_VALUE to remove the key=value pair.
+ * @param {Object} options Parser options.
+ * @param {Boolean} options.write True if writing.
+ * @param {Object} options.state An object to pass to the handler callbacks.
+ * @return {Object|String} The data from the INI file, or the new file content if options.write=true.
+ */
+gpii.iniFile.parse = function (content, handlers, options) {
+    options = Object.assign({}, options);
+
+    var sectionPath = [];
+
+    if (handlers.sectionBegin) {
+        handlers.sectionBegin(options.state, sectionPath);
+    }
+
+    // Called when a [section] has ended, either due to the start of another section, or at the end of file.
+    var endSection = function (depth) {
+        var result = "";
+        // Go back up the stack (if required)
+        while (depth <= sectionPath.length) {
+            // At this point, the sub-sections for the current section are done.
+            if (handlers.sectionEnd) {
+                var endContent = handlers.sectionEnd(options.state, sectionPath);
+                if (options.write && endContent) {
+                    result += endContent;
+                }
+            }
+            if (sectionPath.length === 0) {
+                break;
+            }
+            sectionPath.pop();
+        }
+        return result;
+    };
+
+    var regexFunction = options.write ? XRegExp.replace : XRegExp.forEach;
+
+    var output = regexFunction(content, gpii.iniFile.regex, function (match) {
+        var result = options.write && match.toString();
```

Comment:
Break out the interior of this function "processOneMatch" or so, as well as "endSection", into globals as well

Reply:
Done.

I sometimes disagree with your suggestions, but I'm usually convinced after I follow them.

Reply:
Well sometimes I don't agree with my suggestions either, but I do feel we have become committed to a certain way of working : P